<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Promise</title>
</head>

<body>
</body>

</html>
<script>
  function a1() {

  }
  // get Json
  getJSON(url).then(jsonData => {
    // 回调函数，它会解析得到JSON的值
    // 之后被异步调用 并且接收JSON值作为参考
  })

  const displayUserProfile = profile => {
    console.log(profile)
  }

  getJSON('url').then(res => {
    displayUserProfile(res)
  })

  //  抓取错误
  const handleProfileError = err => {
    console.log(err)
  }
  getJSON('url').then(displayUserProfile, handleProfileError)

  // 另一种写法
  // 语义化更加明显
  getJSON('url').then(displayUserProfile).catch(handleProfileError)

  // 在getJSON、displayUserProfile甚至handleUserProfile抛出的任何异常都会作为参数传到handleUserProfile函数中
  // 等价于
  getJSON('url').then(displayUserProfile((res = null) => {
    // ...
  })).catch(handleProfileError)




  // promise链

  fetch(document)
    .then(response => response.json())
    .then(document => {
      return render(document)
    })
    .then(rendered => {
      cacheInDatabase(rendered)
    })
    .catch(err => {
      handle(error)
    })

  fetch('api/user/profile').then(response => {
    if (response.ok && response.header.get('content-Type') === 'application/json') {
      // do something
    }
  })


  //  幼稚方式
  fetch('api/user/profile').then(response => {
    response.json().then(profile => {
      displayUserProfile(profile)
    })
  })
  //  之所以说这种写法为幼稚方式是因为
  //  promise应该对每次结果的处理是用链式方法
  // 这种方式违背了promise的初衷

  //  该采用的方式
  fetch('api/user/profile').then(response => {
      return response.json()
    })
    .then(profile => {
      displayUserProfile(profile)
    })




  const theURL = 'api/user/profile'
  fetch(theURL)
    .then(callback(1))
    .then(callback(2))
    .then(callback(3))
    .then(callback(4))




  // 解决promise
  function c1(response) {
    let p4 = response.json()
    return p4
  }

  function c2(profile) {
    displayUserProfile(profile)
  }

  let profile_1 = fetch('api/user/profile')
  let profile_2 = profile_1.then(c1)
  let profile_3 = profile_1.then(c2)


  // catch方法实际上就是第一个参数为null的then方法
  // 例如： 
  profile_1.then(null, c)
  profile_1.catch(c)
  // 以上两种书写方式是等价的

  // 相对完整版本的.then 加入了一些出错判断
  // 并且使用了 catch(e) 的 then(null, e) 版本

  fetch('api/user/profile')
    .then(response => {
      if (!response.ok) {
        return null
      }
      let type = response.headers.get('content-type')
      if (type !== 'application/json') {
        throw new TypeError(`Expected JSON, got ${type}`)
      }
      return response.json()
    })
    .then(profile => {
      if (profile) {
        displayUserProfile(profile)
      } else {
        displayUserProfile()
      }
    })
    // .catch(e => {})
    .then(null, e => {
      if (e instanceof NetworkError) {
        displayErrorMessage('check your internet connection')
      } else if (e instanceof TypeError) {
        displayErrorMessage('Something is wrong with your serve')
      } else {
        console.error(e)
      }
    })
</script>